dropshot/src/server.rs

```diff
@@ -543,18 +543,55 @@ impl<C: ServerContext> Service<&TlsConn> for ServerConnectionHandler<C> {
     }
 }
 
-/**
- * A running Dropshot HTTP server.
- */
-pub struct HttpServer<C: ServerContext> {
+/// A running Dropshot HTTP server.
+///
+/// The generic traits represent the following:
+/// - C: Caller-supplied server context
+/// - Closer: Identifies if the server is directly closeable.
+/// - Joiner: Identifies if the server is directly joinable.
+pub struct HttpServer<C: ServerContext, Closer = Closeable, Joiner = Joinable> {
     probe_registration: ProbeRegistration,
     app_state: Arc<DropshotState<C>>,
     local_addr: SocketAddr,
-    join_handle: Option<tokio::task::JoinHandle<Result<(), hyper::Error>>>,
+    closer: Closer,
+    joiner: Joiner,
+}
+
+/// Represents the ability to close a running HTTP server.
+///
+/// Returned from [HttpServer::get_closer].
+pub struct Closeable {
     close_channel: Option<tokio::sync::oneshot::Sender<()>>,
 }
 
-impl<C: ServerContext> HttpServer<C> {
+impl Closeable {
+    /// Closes the server without waiting for it to terminate.
+    ///
+    /// To await termination, see: [Joinable].
+    pub async fn close(mut self) -> Result<(), String> {
+        self.close_channel
+            .take()
+            .expect("cannot close twice")
+            .send(())
+            .expect("failed to send close signal");
+        Ok(())
+    }
+}
+
+/// Identifies that the server cannot be closed directly.
+pub struct Uncloseable {}
+
+/// Represents the ability to join a running HTTP server.
```

Comment:
```suggestion
/// Handle used to wait for a running HTTP server to shut down
///
/// Note that you can obtain this object while the server is still running, in which case `await`ing on this future will block until someone triggers a shutdown _and_ that shutdown completes. 
```

(I didn't like "join" here in the user-facing description because we're a few steps afar from abstractions where that verb is used -- tasks, threads -- and I thought someone might not understand what "join" means here.)